presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddLocalExchanges.java

```diff
@@ -473,19 +480,56 @@ public PlanWithProperties visitTopNRowNumber(TopNRowNumberNode node, StreamPrefe
+                return planAndEnforceChildren(originalTableWriterNode, singleStream(), defaultParallelism(session));
             }
-            else {
-                requiredProperties = singleStream();
-                preferredProperties = defaultParallelism(session);
+
+            if (!isTableWriterMergeOperatorEnabled(session)) {
+                return planAndEnforceChildren(originalTableWriterNode, fixedParallelism(), fixedParallelism());
             }
-            return planAndEnforceChildren(node, requiredProperties, preferredProperties);
+
+            Optional<StatisticAggregations.Parts> statisticAggregations = originalTableWriterNode.getStatisticsAggregation()
+                    .map(aggregations -> aggregations.splitIntoPartialAndIntermediate(variableAllocator, metadata.getFunctionManager()));
+
+            PlanWithProperties tableWriter = planAndEnforceChildren(
+                    new TableWriterNode(
+                            originalTableWriterNode.getId(),
+                            originalTableWriterNode.getSource(),
+                            originalTableWriterNode.getTarget(),
+                            variableAllocator.newVariable("partialrowcount", BIGINT),
+                            variableAllocator.newVariable("partialfragments", VARBINARY),
+                            variableAllocator.newVariable("partialcontext", VARBINARY),
+                            originalTableWriterNode.getColumns(),
+                            originalTableWriterNode.getColumnNames(),
+                            originalTableWriterNode.getPartitioningScheme(),
+                            statisticAggregations.map(StatisticAggregations.Parts::getPartialAggregation)),
+                    fixedParallelism(),
+                    fixedParallelism());
+
+            PlanWithProperties gatheringExchange = deriveProperties(
+                    gatheringExchange(
+                            idAllocator.getNextId(),
+                            LOCAL,
+                            tableWriter.getNode()),
+                    tableWriter.getProperties());
+
+            return deriveProperties(
+                    new TableWriterMergeNode(
+                            idAllocator.getNextId(),
+                            gatheringExchange.getNode(),
+                            originalTableWriterNode.getRowCountVariable(),
+                            originalTableWriterNode.getFragmentVariable(),
+                            originalTableWriterNode.getTableCommitContextVariable(),
+                            statisticAggregations.map(StatisticAggregations.Parts::getFinalOrIntermediateAggregation)),
+                    gatheringExchange.getProperties());
+        }
+
+        @Override
+        public PlanWithProperties visitTableWriteMerge(TableWriterMergeNode node, StreamPreferredProperties context)
+        {
+            throw new IllegalArgumentException("Unexpected TableWriterMergeNode");
```

Comment:
nit: `throw new IllegalStateException("Unexpected node: " + node.getClass().getName());`?

Reply:
It will always be `com.facebook.presto.sql.planner.plan.TableWriterMergeNode`, as nothing extends the `TableWriterMergeNode`

Reply:
Ah okay, I said this because I saw other methods are doing the way it is and thought it might be good to be consistent.